### ISSUE TITLE
non-standard ECDSA sighash types are not preserved, but allowed in PSBT

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#777]
In  https://github.com/rust-bitcoin/rust-bitcoin/pull/669#discussion_r780689066 we decided to allow non-standard sighash types in PSBT partial signatures, on the grounds that the PSBT spec doesn't forbid them and they're allowed by consensus.

However, as Sanket reminds me at https://github.com/rust-bitcoin/rust-miniscript/pull/290#discussion_r783589004, our `EcdsaSighHashType` does not actually distinguish between different nonstandard sighash types, and effectively silently converts them all to 1. This will cause bugs in PSBT if somebody actually tries to use the sighash type.

Solutions are:
* Petition the Bitcoin BIPs repo to restrict PSBT to disallow nonstandard sighash types
* Restrict sighash types in *our* code, regardless of what the BIP says
* Fix the `EcdsaSigHashType` to keep track of which nonstandard type it uses (and then do the right thing during sighash computations, serialization, etc)

**Comment by apoelstra** [comment]
Ah, I see Sanket opened https://github.com/rust-bitcoin/rust-bitcoin/pull/776/ which implements my second suggestion.

**Comment by Kixunil** [comment]
I think it'd be great to at least probe other Bitcoin devs what they think about restricting things. I strongly dislike doing things against standards, especially in Bitcoin. Historically breaking standard caused us pain.

**Comment by sanket1729** [comment]
@Kixunil, we are already doing the standardness check in the library for one component. In some sense, we are already reliant on it.  
https://github.com/rust-bitcoin/rust-bitcoin/blob/907b3a7a6a1c8d275a911b45c3931a2dcc5cd6db/src/util/psbt/serialize.rs#L174

Most of the reason the psbt standard has sighash as u32 instead of allowed values is that it is how it is serialized in segwitv0 signatures. 
A few reasons to restrict SigHashType to standard values instead of free-floating u32. 
- Non-standard transactions are not relayed by the network, but they can be mined. We should stand along the side of warning users that they are dealing with non-standard transactions, instead of succeeding the API. Bitcoin Core, or BIP does not mention this, but I think there is a lot of value in doing this and we get cleaner types. 
- There are virtually zero users of non-standard sig hash types. If they want to sign a transaction, they can use the raw APIs directly that supports non-standard sighash types. I don't think we should sacrifice high-level psbt APIs for something that will never be used. 
- The hairy details of which sighash types are allowed are even more complicated. AFAIR, bits 2 and 3 are free-flow bits, and there is some weird masking going on in `from_u32_consensus`. 
- The Bip also does not mention that bitcoin publickeys should be valid, it merely states that they should be 33/65 bytes. I think there may even be some more ECDSA standardness signature rules that we are restricting. By nature of a strong type system in rust, we benefit a lot from restricting these to only valid values that a struct can take.

Overall, I think our approach should be:
-  We should have support to parse, accept and validate blocks, transactions, and signatures containing non-standard data. We want to support validation and use-cases like block explorers that might potentially have to deal with non-standard transactions.   In short, all code that deals with data over the bitcoin p2p network should follow consensus rules and support non-standard stuff.
-  All of the application-level user APIs(like psbt) should treat non-standard stuff as second-class citizens and optimize the code quality for standardness rules. We should not allow users to create pbsts that might be non-standard when spent.

**Comment by Kixunil** [comment]
Great, at least we have a strong rationale for this and not blindly ignoring it nor we overlooked it. :)
I really want to agree with you it just feels a bit like a minefield. So let's explore:

* I can't imagine any security vulnerability stemming from making things stricter except for attacks where two pieces of software disagree on validity of data. I'm not sure what a good mitigation would be here. "Write better doc" sounds too hand-wavy. However nobody I know of builds "transaction firewalls" today so that at least buys us some time. Also I don't think there's anything stricter than this so having rust-bitcoin at the boundary should be fine.
* Someone could get annoyed we don't support the features he wants. This is a more general problem, we can tell people to make a PR or fork it.
* Communication issues - someone could falsely rely on thinking this is according to the standard. I suggest to very clearly document this is not true BIP 174 but a stricter form. I was thinking calling it "stricter PSBT" in docs
* Reputation issue - someone may be worried that we don't follow the standard perfectly (a reasonable caution I'd say). This library really doesn't have a competition if someone wants Rust and we don't get paid by consumers so probably not a big deal. Also anyone reasonable would read the rationale and at least see we're not ignorant, stupid, careless...
* Probably all of these can be solved by updating the BIP or writing a new one to describe the strictness.

Overall doesn't seem too bad now that I look at it but would be interested if someone has more experience with (defense agains) attacks I mentioned and still would like to document it well.

**Comment by sanket1729** [comment]
> Probably all of these can be solved by updating the BIP or writing a new one to describe the strictness.
I will bring this up with the author of the BIP.

**Comment by dr-orlovsky** [comment]
Is this issue fully or only partially covered by #779?

**Comment by apoelstra** [comment]
It's not really related. #779 will preserve hashtypes when reserializing PSBTs but when you actually try to use them (e.g. with the rust-miniscript finalizer) then the `PsbtSigHashType` will need to be converted to a "real" `SigHashType` and then the information will be lost.

**Comment by sanket1729** [comment]
This also affects the `signature_hash` computation algorithm and is thus not related to the psbt spec completely. After some deliberation on the issue. I think we should just the bullet and add another enum variant to `EcdsaSigHashType`.

**Comment by Kixunil** [comment]
Hmm, this is in 0.28 milestone but not closed, kinda strange isn't it?

**Comment by tcharding** [comment]
Since we now use `PsbtSigHashType` for the `sighash_type` field of `Input` (in psbt), is this issue resolved? Users can store an non-standard sighash type and get at it with the current API. The value will only be lost (as described in the issue description) when converted to an `EcdsaSigHashType`.

**Comment by tcharding** [comment]
Do we need this codeseparator/non-standard sighash type stuff in the taproot release? This issue is marked, as noted above, with the 0.28 milestone - can we remove that? Tracking issue #887 shows that this work is complicated enough that if we really want it in the Taproot release it is going to block the release.

**Comment by apoelstra** [comment]
@tcharding we definitely do not need codeseparator stuff for the Taproot release. For non-standard sighash stuff I defer to @sanket1729 about how important the remaining issues are to him.

**Comment by sanket1729** [comment]
I think #796 is all I need that is corrected tagged with RC-fix. 

As far as non-standard stuff goes, I think this is really really a low priority because no one ever uses it.  Application layer protocols like psbt do not support things like code_sep because it adds complications for no reason. 

In my opinion:

- I think the current API, with the improvements from #796 should be enough for all purposes. 
- We should clearly identify it and error on non-standard things, but do not any invest effort making first-class API support for it unless someone complains and address this on case-by-case/demand basis.
- The only use-case for non-standard stuff is block-explorers who want to display information about the transactions/Blocks. Considering this, we should double-check whether our Encodable/Decodable traits correctly parse non-standard things.
- As far as API support, we should document that we do not support verification of non-standard signatures. We review all of our APIs with standardness rules clearly document what we don't support.

**Comment by tcharding** [comment]
Cool, cheers @sanket1729, so we can close this issue, right?

**Comment by tcharding** [comment]
I'm removing the 0.28 milestone - commenting here so its super explicit and anyone can re-add it if I'm incorrect.

**Comment by sanket1729** [comment]
@tcharding, Should we change this issue to document non-standard support for the things that we don't support? Or perhaps open another issue for it? 

As you correctly observe in #887, implementing full non-standard support is complicated. I would like to use that issue to document that we don't support it.  I think signature data structures `EcdsaSig` and `SchnorrSig` after #796 clearly document non-standard things.

Here is a suggestion:
- The algorithms for `legacy_signature_hash`, `segwit_signature_hash` should have comments that highlight `OP_CODESEP` are not supported. I think `taproot_signature_hash` already does the correct thing here.
- In `legacy_signature_hash` we should rename the `script_pubkey` parameter to `script_code`. Add a comment stating that `script_pubkey` is `script_code` if the script contains no `op_codesep`. Optionally, add a comment that users using op_codesep can compute the script_code according to `FindAndDelete` (add a link to bitcoin core `FindAndDelete`)
- In `segwit_signature_hash`, similarly, add a comment about `script_code` calculation linking to BIP143 issue. We can either link to BIP143 for `script_code` definition or also highlight the three cases from the BIP:

> For P2WPKH witness program, the scriptCode is 0x1976a914{20-byte-pubkey-hash}88ac.
For P2WSH witness program,
if the witnessScript does not contain any OP_CODESEPARATOR, the scriptCode is the witnessScript serialized as scripts inside CTxOut.
if the witnessScript contains any OP_CODESEPARATOR, the scriptCode is the witness script but removing everything up to and including the last executed OP_CODESEPARATOR before the signature checking opcode being executed, serialized as scripts inside CTxOut. (The exact semantics is demonstrated in the examples below). 

Slightly related to this: I had to do something awkward downstream: https://github.com/rust-bitcoin/rust-miniscript/blob/40cc0c917857953cdf187d916ade1e9005a6dad7/src/psbt/mod.rs#L865-L874. I think a rust-bitcoin API to get `script_code` from `WPubkeyHash` would be valuable. After we implement this, we can link this in the documentation of `segwit_signature_hash` so that users of `p2wpkh` (the most common output type) can benefit from it.

**Comment by sanket1729** [comment]
@tcharding sorry about the closed PRs :( . We can wait some time to see if others have opinions on it. I will try to be more active in responding to new PRs with concepts ACKs/NACKs in order to avoid development time before its too late.

**Comment by tcharding** [comment]
No need to burden yourself on my account, we should optimize for reviewer time not developer time. I learned a bunch from that work, there was some concerted effort digging into bitcoin core which well over due for me.

**Comment by tcharding** [comment]
I will process your suggestion and work up a PR, leave it with me. Can you answer me one thing, you mention `FindAndDelete` but the serialization code in bitcoin core never calls `FindAndDelete` on the script? AFAICT its only called to do verification an we call through to `rust-bitocinconsensus` to do verification. The serialization code removes all instances of `OP_CODESEPARATOR` in `interpreter.cpp`, or am I missing something? No rush, just if you have time to look.

**Comment by jrakibi** [comment]
Can this be closed now? We already support non-standard sighash types.
